README.md
# Native [V-lang](https://vlang.io) implementation of [Rosie-RPL](https://rosie-lang.org/)

[Rosie](https://rosie-lang.org/) is a pattern language (RPL for short). A little bit like
regex, but aiming to solve several of the [regex issues](https://jamiejennings.com/posts/2021-09-23-dont-look-back-2/).
All credits to Jamie A. Jennings and her friends for this job.

This project (native V-lang implementation of RPL) is work in progress (beta), but ready to be tested
in the field. APIs may still change, CLI is available, and a REPL is on the todo list.
The current version is fully functional: it parses and compiles all files in Jamie's RPL libary (./rpl), including
the rpl files to parse RPL code, and it successfully executes all (inline) unittests in this folder.

Very similar to a compiler, the project consists of the following modules:
- A core 0 parser, written in V, which is able to parse rpl input into an AST
- A RPL-parser which uses the core-0 parser to create the byte-code for the RPL-parser
- An Expander (and optimizer) that expands macros and aliases
- A compiler backend, which converts the AST into virtual machine byte code
- A virtual machine runtime, able to execute the byte code instructions and match input against the pattern
- A CLI module (command line interface)
- A unittest module, which support RPL inline tests to valid patterns in '*.rpl' files
- A disassembler that prints the byte code instructions generated for a specific pattern
- A tracer utility (via CLI) that greatly helps with debugging input against pattern

Not yet available:
- Possibly an additional compiler backend that generates native V code
- A shared library and language integration, e.g. Python


## Project objectives

- Be compliant with the RPL Language Reference
- Easy and intuitive to use in V-lang projects
- A REPL to test and debug rpl pattern easily
- Jamie's implementation has nice support for grep-like search, colored output, and also tree-like output
  to review details of the AST. I'd like to reach at least a similar level of user support.
- Integration with other lanugages such as Python, Julia, C/C++, Rust, Java, JavaScript, etc.
  The more popular languages are supported, the better.  (I wish V-lang would have a python integration module)
- A Visual-Code Studio plugin would be nice. Syntax highligting for rpl files, readonly view of
  disassembled rplx files, compile rpl files upon save or manually triggered, automatically run
  unittest, etc..


## A bit of history

The project started with a tiny virtual machine (v1), able to load and execute '\*.rplx' files
(compiled RPL code), generated by Rosie's original compiler. It is working, but is not battle tested.
By now, the virtual machine has evolved (v2) and is no longer backwards compatible. We are
still able to read and execute '\*.rplx' files, but we'll not put more effort into it.

Please note that neither the '\*.rplx' file structure nor the byte codes of the virtual
machine are part of Rosie's specification and thus are subject to change without
formal notice from the Rosie team.

Originally the project was a proof-of-concept aiming at getting pratical experience with V
and validate it's promises. I decided to use Rosie because I like many of it's ideas, and thought
it would be a good contributions to V as well.

Obviously I had to start somewhere, and I decided to start with the RPL runtime. The original
RPL runtime is written in C, whereas the compiler and frontend is a mixture of C and Lua.
The V implementation started as copy of the C-code, gradually introducing more and more V constructs,
and also replacing 'unsafe' pointer arithmetics. V's C-to-V translator was not yet available,
hence I translated and reengineered the code manually.

Next I've added an RPL parser written in V, able to read and parse RPL source code into an
AST (intermediate representation). It successfully reads all '\*.rpl' files provided in Rosie's library,
including the rpl files implementing the RPL language specification itself.

And then a compiler that generates RPL-VM byte code instructions (v2). Now I had all core components
required available and fully implemented in V-lang.

Performance tests and optimisations, a proper CLI, and the ability to easily plug-in additional parsers,
optimizers and compilers, were now top of my prio list.

First I've added a benchmark module for the runtime (matching input against a pattern), which let to a greatly
improved runtime performance. It was a good learning excercise for me on how certain V features affect
the performance, but also which ones are left to the C-Compiler and how the CPU architecture affects the results.
But I'm certainly not an X86 or SIMD assembler experts, neither a CPU profiling expert.

Slowly the project is moving into a more stable mode, evident by the enhancements that followed:
- A CLI (with colored output)
- A tracer (debugger) to more easily analyse what is happening when matching input against a pattern
- An Engine, that allows to more easily plug-in different versions of parsers, optimizer and compilers.
- Tons of entries in todo.md and TODO comments in the source code

As mentioned, this project started as PoC to practically test and gain some experience with V-lang.
Despite some rough edges here and there, so far I'm mostly pleased. See
[here](https://github.com/jdonnerstag/vlang-lessons-learnt/wiki) for my very own FAQ and
"things to remember" list. I find the V-code much easier to read and maintain
then comparable C-code. Compiler speed is definitely a plus as well, allowing for quick code-test cycles.
Occassionaly I wish a V-interpreter or -debugger would already be available, to help me find and fix
issues. For now, adding and removing debug messages is what I do (and why V built time is so important).


## CMD, PS, bash etc.. and the problem with quotes

This project can be embedded in other V projects, but it also comes with a cli. The cli has subcommands
such as 'grep' and 'match', expecting a pattern argument such as `"a" ~ "b"`. The pattern has
double quotes and spaces. Both are treated differently, depending on your shell (bash, CMD, PS, ...).
Because I stumbled upon it more then ones, I've collected links to blogs that helped me understand
[here](https://github.com/jdonnerstag/vlang-lessons-learnt/wiki/Command-lines-and-how-they-handle-single-and-double-quotes).


## Differences with Jamie's implementation

I've tried to limit differences as much as possible, but ocassionally and very conciously, I've decided to differ.

- built-in overrides: I've added a 'builtin' binding attribute in addition to 'alias' and 'local', so that e.g.
  `builtin alias ~ = [:space:]+` will override the builtin implementation, and it'll be applied to all
  patterns, including the imported packages, and their imports.
- I did not implement '#'. It's not used in any of Jamie's rpl files, which is a good indication, that it is not needed.
- The cli commands and outputs are slightly different
   - Added support to print the byte-code (diassembler)
   - The tracing output looks completely different, IMHO more concise and better readable.
   - The supported rcfile variables are different, also 'add_xxx' allows to add a libpath or color.
- Because performance analysis revealed that captures occassionaly take significant time (%) off the end-to-end
  processing time. The user function to execute a match allows to provide a list of bindings, which are
  really needed, superseding what is defined in the rpl files.
- As alluded to above, my byte codes have evolved quite a bit, significantly contributing to the runtime performance.
- In RPL 1.x the &-operator is equivalent to {>p q}. Which IMHO is misleading, everybody beliefs its concatenation,
  and I've not seen it being used anywhere in the lib files. Hence, we do not support it. I think Jamie plans
  to remove it in RPL 2.x as well.
- Several improvements on Jamie's todo lists, have been implemented in my project, such as
    - multiple byte-code entrypoints (per file)
	- re-usable code-blocks (aka functions; but w/o parameters) reducing the file size many times
	- Some commonly used (and complex) builtin pattern, e.g. "." and "~" have their own byte-code instruction
	  with performance optimized V-code implementations.
- Based on the experience gained throughout the project, I made a couple of suggestions to Jamie on how to evolve
  the Rosie Pattern language in version. I've started with my own one in "./rpl/rosie/rpl_3_0_jdo.rpl". So far,
  it is only thoughts and not yet implemented anywhere.


CHANGELOG.md
## V 0.1.10
- Started with CLI
- Added rc-file (~/.rosierc) support
- Started with a benchmark module
- Introduced byte code instructions for ".", "~", [:digit:] for better performance, and more readable byte code.
- Introduced byte code instructions "until_char" and "until_set" to speed up "find", "keepto" and "findall" macros
- Compiler supports manually overriding which variables are captured (to improve performance)
- Support for global builtin overrides
- Support for multiple entry-points in the byte code
- Started with an Engine to plug & play with the different parser and compiler versions available
- Moved Expander into a separate module
- Support pre-compiled *.rplx files (e.g. ./rpl/rosie/rpl_1_3_jdo.rplx), which is now the default for the RPL parser itself
- Added a new parser for RPL-3.0, which is my attempt to bring the pattern language to the next level
- Added a MasterParser which is able to switch automatically between the different parsers, depending on the rpl
  language version. This way, the RPL-3.x parser is able to import RPL-1.x files from the library.
- Added a 'halt' macro which yields a 'halt' byte code instruction, and captures the client-pattern. It also
  allows to continue parsing afterwards. This is intended for 'rpl 1.3' and line-based inputs (e.g. CSV files)
- Added a quote macro and byte-code instruction, because it is such a common case
- Added an until macro, which e.g. allows to move forward to the next line (or eof). This is nice for
  comments, rest-of-line (syntax error), or CSV files.
- Added few more meta-data to the rplx file
- The RPL-1.3 parser byte code is now loaded via $embed_file(), so that in prod mode it doesn't need to
  be loaded from file. Now the exe is fully self-contained.
- Added -show_timings option to some CLI subcomands (e.g. compile)

## V 0.1.9
- Fixed issues with "(a)+" like pattern. See https://gitlab.com/rosie-pattern-language/rosie/-/issues/123
- Clarified back-reference resolution process. See https://gitlab.com/rosie-pattern-language/rosie/-/issues/121
- Complete rebuild of disjunctions, such as `[..]` and `[^ ..]`

## V 0.1.8
- This is the first version, that passes all original Rosie RPL (inline) unittests (see ./rpl)
- Added capture callback function, to allow clients to receive capture "streams"
- Added support for VM function calls with ok/err support. Which is now the default for word-boundary (~).
   Theoretically that should allow multiple entry points as well, but hasn't been tested yet.

## V 0.1.5
- We now have a core 0 parser, byte code compiler, virtual runtime and disassembler
- predicate, multiplieres, word-boundary, choices, eof and bof are implemented and seem to be working.
  There are numerous tests in each package.

## V 0.1.0
- So far this is merely a proof-of-concept
- The virtual machine (runtime) seems to be working, but testing is still a bit akward. It is alpha may be
- Since generating the byte-code is a bit akward, I looked at the compile backend, but that prooved difficult
- So started with a core 0 parser. Written in V-lang, it is not yet leveraging the runtime. The objective is to use
  the runtime though. Hence the core 0 parser is not written with correctness in mind, and error handling is
  probably a bit weak as well.

rosie_cli.v
module main

// This is rosie's command line interface (cli). Please see 'v . --help' for the full list
// of options and subcommands.

import v.vmod
import cli
import os
import rosie.rcli

fn main() {
	vm := vmod.decode(@VMOD_FILE) ?

	mut app := cli.Command{
		name: vm.name
		description: vm.description
		version: vm.version
		posix_mode: true
		flags: [
			cli.Flag{	// TODO not sure how to do -vvvv or -v 10, where 10 is optional
				flag: .bool
				name: 'verbose'
				abbrev: 'v'
				description: 'Output additional messages'
			},
			cli.Flag{
				flag: .string
				name: 'rpl'
				abbrev: ''
				description: 'inline RPL statements'
			},
			cli.Flag{
				flag: .string
				name: 'file'
				abbrev: 'f'
				description: 'Load a RPL file'
			},
			cli.Flag{
				flag: .bool
				name: 'norcfile'
				abbrev: ''
				description: 'Skip initialization file'
			},
			cli.Flag{
				flag: .string
				name: 'rcfile'
				abbrev: ''
				description: 'Explicitly define the initialization file'
			},
			cli.Flag{
				flag: .string
				name: 'libpath'
				abbrev: ''
				description: 'Directories to search for rpl modules'
			},
			cli.Flag{
				flag: .string
				name: 'colors'
				abbrev: ''
				description: 'Color/pattern assignments for color output'
			},
		]
		execute: fn (cmd cli.Command) ? {
			cmd.execute_help()
		}
		commands: [
			cli.Command{
				name: 'config'
				description: 'Print rosie configuration information'
				posix_mode: true
				execute: rcli.cmd_config
				flags: [
					cli.Flag{
						flag: .bool
						name: 'lib'
						abbrev: 'l'
						description: "List the config as if in (shared) library mode. Default: cli mode."
					},
					cli.Flag{
						flag: .bool
						name: 'color_test'
						description: "Additional print a line for each color defined"
					},
				]
			},
			cli.Command{
				name: 'list'
				description: 'List patterns, packages, and macros'
				posix_mode: true
				required_args: 0
				flags: [
					cli.Flag{
						flag: .string
						name: 'filter'
						abbrev: 'f'
						description: "List all names that have substring 'filter' (default: *)"
					},
				]
				execute: rcli.cmd_list
			},
			cli.Command{
				name: 'grep'
				description: 'In the style of Unix grep, match the pattern anywhere in each input line'
				required_args: 1
				usage: '<pattern> [<filename>] [<filename>] ...'
				posix_mode: true
				flags: grep_match_flags
				execute: rcli.cmd_grep
			},
			cli.Command{
				name: 'match'
				description: 'Match the given RPL pattern against the input'
				required_args: 1
				usage: '<pattern> [<filename>] [<filename>] ...'
				posix_mode: true
				flags: grep_match_flags
				execute: rcli.cmd_match
			},
			/*
			cli.Command {
                name: 'repl'
                description: 'Start the read-eval-print loop for interactive pattern development and debugging'
                posix_mode: true
                execute: fn (cmd cli.Command) ? {
                   println('repl subcommand')
                   return
                }
            },
			*/
			cli.Command{
				name: 'test'
				description: 'Execute pattern tests written within the target rpl file(s)'
				usage: '<filenames> [<filenames>] ...'
				posix_mode: true
				required_args: 1
				execute: rcli.cmd_test
				flags: [
					cli.Flag{
						flag: .bool
						name: 'show_timings'
						abbrev: ''
						description: "Output additional timing information per execution phase"
					},
				]
			},
			cli.Command{
				name: 'expand'
				description: 'Expand an rpl expression to see the input to the rpl compiler'
				posix_mode: true
				usage: '<expression>'
				required_args: 1
				execute: rcli.cmd_expand
			},
			cli.Command{
				name: 'disassemble'
				description: 'Print the virtual machine byte code instructions for the <expression>'
				posix_mode: true
				usage: '<expression or file path>'
				required_args: 1
				execute: rcli.cmd_disassemble
			},
			cli.Command{
				name: 'compile'
				description: 'Compile one or more patterns into a rplx file'
				posix_mode: true
				usage: '<expression or rpl-file> [<entrypoint> ..]'
				required_args: 2
				execute: rcli.cmd_compile
				flags: [
					cli.Flag{
						flag: .string
						name: 'output'
						abbrev: 'o'
						description: "Output rplx file name or module directory (V output)"
					},
					cli.Flag{
						flag: .string
						name: 'language'
						abbrev: 'l'
						description: "Select a specific language, e.g. stage_0, 1.3, 3.0"
					},
					cli.Flag{
						flag: .string
						name: 'compiler'
						abbrev: 'c'
						description: "Select the compiler, e.g. vm_v2 (default), vlang"
					},
					cli.Flag{
						flag: .bool
						name: 'show_timings'
						abbrev: ''
						description: "Output additional timing information per execution phase"
					},
				]
			},
		]
	}
	app.setup()
	app.parse(os.args)
}

const grep_match_flags = [
	cli.Flag{
		flag: .string		// TODO we are not yet supporting all these formats, and I'm not sure they all provide real value
		name: 'output'
		abbrev: 'o'
		description: 'Output style, one of: jsonpp, color, data, bool, subs, byte, json, line'
	},
	cli.Flag{
		flag: .bool
		name: 'wholefile'
		abbrev: 'f'
		description: 'Read the whole input file as single string (default: line by line)'
	},
	cli.Flag{
		flag: .bool
		name: 'all'
		abbrev: 'a'
		description: 'Output all lines, including the non-matching lines'
	},
	cli.Flag{
		flag: .bool
		name: 'fixed-strings'
		abbrev: 'F'
		description: 'Interpret the pattern as a fixed string, not an RPL pattern'
	},
	cli.Flag{
		flag: .bool
		name: 'time'
		abbrev: ''
		description: 'Time each match, writing to stderr after each output'
	},
	cli.Flag{
		flag: .bool
		name: 'profile'
		abbrev: ''
		description: 'Print instruction execution statistics (requires to compile source code with -cg)'
	},
	cli.Flag{
		flag: .bool
		name: 'trace'
		abbrev: 'c'
		description: 'Pretty print all captures that matched'
	},
	cli.Flag{
		flag: .bool
		name: 'unmatched'
		abbrev: 'u'
		description: 'Also print captures that did not match (requires --trace)'
	},
	cli.Flag{
		flag: .bool
		name: 'incl_aliases'
		abbrev: 'i'
		description: 'Enable capture also for aliases (requires --trace)'
	},
	cli.Flag{
		flag: .bool
		name: 'show_timings'
		abbrev: ''
		description: "Output additional timing information per execution phase"
	},
]
